src/remap/function/compact.rs

```diff
@@ -0,0 +1,406 @@
+            _ => false,
+        }
+    }
+}
+
+impl Expression for CompactFn {
+    fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        let options = CompactOptions {
+            recursive: match &self.recursive {
+                Some(expr) => expr.execute(state, object)?.try_boolean()?,
+                None => false,
+            },
+
+            null: match &self.null {
+                Some(expr) => expr.execute(state, object)?.try_boolean()?,
+                None => false,
+            },
+
+            string: match &self.string {
+                Some(expr) => expr.execute(state, object)?.try_boolean()?,
+                None => false,
+            },
+
+            map: match &self.map {
+                Some(expr) => expr.execute(state, object)?.try_boolean()?,
+                None => false,
+            },
+
+            array: match &self.array {
+                Some(expr) => expr.execute(state, object)?.try_boolean()?,
+                None => false,
+            },
+        };
+
+        match self.value.execute(state, object)? {
+            Value::Map(map) => Ok(Value::from(compact_map(map, &options))),
+            Value::Array(arr) => Ok(Value::from(compact_array(arr, &options))),
+            _ => unreachable!(),
+        }
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        self.value
+            .type_def(state)
+            .fallible_unless(value::Kind::Map | value::Kind::Array)
+            .with_constraint(value::Kind::Map | value::Kind::Array)
+    }
+}
+
+/// Compact the value if we are recursing - otherwise, just return the value untouched.
+fn recurse_compact(value: Value, options: &CompactOptions) -> Value {
+    match value {
+        Value::Array(array) if options.recursive => Value::from(compact_array(array, options)),
+        Value::Map(map) if options.recursive => Value::from(compact_map(map, options)),
+        _ => value,
+    }
+}
+
+fn compact_map(map: BTreeMap<String, Value>, options: &CompactOptions) -> BTreeMap<String, Value> {
+    map.into_iter()
+        .filter_map(|(key, value)| {
+            let value = recurse_compact(value, options);
+            if options.is_empty(&value) {
+                None
+            } else {
+                Some((key, value))
+            }
+        })
+        .collect()
+}
+
+fn compact_array(array: Vec<Value>, options: &CompactOptions) -> Vec<Value> {
+    array
+        .into_iter()
+        .filter_map(|value| {
+            let value = recurse_compact(value, options);
+            if options.is_empty(&value) {
+                None
+            } else {
+                Some(value)
+            }
+        })
+        .collect()
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::map;
+
+    #[test]
+    fn test_compacted_array() {
+        let cases = vec![
+            (
+                vec![Value::from(1), Value::from(2)],              // expected
```

Comment:
minor nit: these can be written as `1.into()`, etc, which makes the tests easier to parse as a reader.